patches/tModLoader/Terraria/ModLoader/ContentCache.cs

```diff
@@ -0,0 +1,93 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+
+namespace Terraria.ModLoader;
+
+/// <summary>
+/// Container object for <see cref="ILoadable"/> instances added for a <see cref="Mod"/>
+/// </summary>
+internal class ContentCache
+{
+	private static event Action OnUnload;
+	private static Dictionary<nint, IList> _cachedContentForAllMods = new();
+
+	internal static bool hasLoadingStarted;
+	internal static bool loadingContent;
+
+	internal static void Unload() {
+		Interlocked.Exchange(ref OnUnload, null)?.Invoke();
+	}
+
+	private static void UnloadStaticCache() {
+		_cachedContentForAllMods.Clear();
+	}
+
+	public static IEnumerable<T> GetContentForAllMods<T>() where T : ILoadable {
+		if (!hasLoadingStarted || loadingContent) {
+			// Content has not fully loaded yet, so we can't rely on the cache.
+			// Return a lazy enumerable instead.
+			return ModLoader.Mods.SelectMany(static m => m.GetContent<T>());
+		}
+
+		// Check of the cache already exists
+		// It will already be a ReadOnlyList<T>, so it just needs to be cast back to it
+		nint handle = typeof(T).TypeHandle.Value;
+		if (_cachedContentForAllMods.TryGetValue(handle, out IList cachedContent))
+			return (IReadOnlyList<T>)cachedContent;
+
+		// Construct the cache
+		if (_cachedContentForAllMods.Count == 0)
+			OnUnload += UnloadStaticCache;
```

Comment:
No need for the `OnUnload` `event` here now. Since you don't have any static fields on generic types.